firmware/application/apps/ui_recon.cpp

```diff
@@ -124,150 +128,87 @@ void ReconView::colorize_waits() {
-        }
+
+    // For ranges, save the current frequency instead.
+    if (entry.type == freqman_type::Range) {
+        entry.frequency_a = freq;
+        entry.frequency_b = 0;
+        entry.type = freqman_type::Single;
+        entry.modulation = last_entry.modulation;
+        entry.bandwidth = last_entry.bandwidth;
+        entry.step = freqman_invalid_index;
     }
+
+    auto it = freq_db.find_entry(entry);
+    auto found = (it != freq_db.end());
+
+    if (found && warn_if_exists)
+        nav_.display_modal("Error", "Frequency already exists");
+
+    if (!found)
+        freq_db.append_entry(entry);
+
     return true;
 }
 
 bool ReconView::recon_load_config_from_sd() {
-    File settings_file{};
-    size_t length{0};
-    size_t file_position{0};
-    char* pos{NULL};
-    char* line_start{NULL};
-    char* line_end{NULL};
-    char file_data[257]{};
-    uint32_t it{0};
-    uint32_t nb_params{RECON_SETTINGS_NB_PARAMS};
-    std::string params[RECON_SETTINGS_NB_PARAMS]{};
+    /* NB: This function is only called once and the defaults
+     * are already set in the class initializers. */
+    // TODO: remove?
+    // // Assign default values.
+    // input_file = "RECON";
+    // output_file = "RECON_RESULTS";
+    // recon_lock_duration = RECON_MIN_LOCK_DURATION;
+    // recon_lock_nb_match = RECON_DEF_NB_MATCH;
+    // squelch = RECON_DEF_SQUELCH;
+    // recon_match_mode = RECON_MATCH_CONTINUOUS;
+    // wait = RECON_DEF_WAIT_DURATION;
 
     make_new_directory(u"SETTINGS");
 
-    auto result = settings_file.open(RECON_CFG_FILE);
-    if (!result.is_valid()) {
-        while (it < nb_params) {
-            // Read a 256 bytes block from file
-            settings_file.seek(file_position);
-            memset(file_data, 0, 257);
-            auto read_size = settings_file.read(file_data, 256);
-            if (read_size.is_error())
-                break;
-            file_position += 256;
-            // Reset line_start to beginning of buffer
-            line_start = file_data;
-            pos = line_start;
-            while ((line_end = strstr(line_start, "\x0A"))) {
-                length = line_end - line_start - 1;
-                params[it] = std::string(pos, length);
-                it++;
-                line_start = line_end + 1;
-                pos = line_start;
-                if (line_start - file_data >= 256)
-                    break;
-                if (it >= nb_params)
-                    break;
-            }
-            if (read_size.value() != 256)
-                break;  // End of file
+    File settings_file;
+    auto error = settings_file.open(RECON_CFG_FILE);
+    if (error)
+        return false;
 
-            // Restart at beginning of last incomplete line
-            file_position -= (file_data + 256 - line_start);
-        }
-    }
+    auto line_nb = 0;
+    auto reader = FileLineReader(settings_file);
+    for (const auto& line : reader) {
+        if (line_nb == 0)
```

Comment:
Just FYI, you could save another 50 bytes by changing all these ifs to a switch statement.

    for (const auto& line : reader) {
        switch (line_nb) {
            case 0: input_file = trim(line); break;
            case 1: output_file = trim(line); break;
            case 2: parse_int(line, recon_lock_duration); break;
            case 3: parse_int(line, recon_lock_nb_match); break;
            case 4: parse_int(line, squelch); break;
            case 5: parse_int(line, recon_match_mode); break;
            case 6: parse_int(line, recon_match_mode); break;
            default: break;
        }
        line_nb++;
    }

Reply:
I had written that first, then I took it back out because I wanted to be able to break early from the loop in the case that we exceed line count. I really wish the language supported breaking out a scope. I just added a bool and another check.

Reply:
Does't c++ support "goto"?  :-)   I thought about that too, but figured it didn't matter if the config file had extra lines that are read then ignored (probably just means it was written with a future version of FW), so no need to break out of the outer loop.